imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -368,6 +382,101 @@ def set_spin_cycle(
     return l1b_de
 
 
+# TODO: The spin cycle function above needs to be updated for DEs. We cannot assume
+#  there are 28 spins per ASC and we should calculate the spin start number based on the
+#  corresponding L1A spin data Acq Start for the ASC. The implementation below should be
+#  should be used for the DE rather than the above function, but in the interest of time
+#  the below function is only hooked up to the histogram rates processing and should be
+#  integrated into the DE processing in a later PR.
+# TODO: Break up the invalid spin ASC removal and the code to find the closest DE/Hist
+#  and spin ASCs into their own functions.
+def set_spin_cycle_from_spin_data(
+    l1a_science: xr.Dataset, l1b_science: xr.Dataset, spin_data: xr.Dataset
+) -> xr.Dataset:
+    """
+    Set the spin cycle for each direct event using the L1A spin data.
+
+    Parameters
+    ----------
+    l1a_science : xr.Dataset
+        The L1A Histogram or Direct Event dataset.
+    l1b_science : xr.Dataset
+        The L1B Histogram Rate or Direct Event dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+
+    Returns
+    -------
+    l1b_science : xr.Dataset
+        The L1B science dataset with the spin cycle added for each direct event.
+    """
+    acq_start, _acq_end = convert_start_end_acq_times(spin_data)
+
+    spin_met_per_asc = spin_data["shcoarse"].values
+    science_met_per_asc = ttj2000ns_to_met(l1a_science["epoch"])
+
+    # Find the closest start_acq for each direct event
+    # computes the index of the closest spin_met_per_asc for each science_met_per_asc
+    # so the resulting array will be of length len(science_met_per_asc), one index per
+    # ASC, but the value of each index will be the index of the closest spin data.
+    closest_start_acq_indices = np.abs(
+        science_met_per_asc[:, None] - spin_met_per_asc
+    ).argmin(axis=1)
+
+    valid_asc_mask = []
+    for spin_idx in closest_start_acq_indices:
+        spin_count = spin_data["num_completed"].isel(epoch=spin_idx)
+        valid_asc_mask.append(spin_count >= 28)
```

Comment:
I think you can avoid the loop here with numpy indexing.
```suggestion
    # It isn't a valid cycle if we didn't complete all 28 spins in an ASC
    valid_cycles = spin_data["num_completed"] == 28
    valid_asc_mask = valid_cycles.values[closest_start_acq_indices]
```